POC2: Delay deletion of __name__ label (bool flags)

jcreixell opened the pull request:
- Delay dropping the  `__name__` label to the end of the query evaluation. This allows optionally preserving it via the `label_replace` and `label_join` functions, and helps prevent the dreaded "vector cannot contain metrics with the same labelset" error.
  - This implementation extends the `Series` and `Sample` structs with a boolean flag indicating whether the `__name__` label should be deleted at the end of the query evaluation.
    - The advantage of this implementation is that it is more transparent to the user; there is no way to externally inject labels that change the behavior of a query.
    - The downsides are:
      - It is more intrusive, as it requires extending existing structs with meta information, and tweaking multiple places in the query engine.
      - And more importantly, propagation of the name-dropping information happens manually, and there is a higher risk of "forgetting" to do so in certain places.
  -  The `label_replace` and `label_join` functions can still access the value of the `__name__` label, even if it has been previously marked for deletion. If  `__name__` is used as target label, it won't be dropped at the end of the query evaluation.
  - This code passes all existing tests.
  - See https://github.com/jcreixell/prometheus/pull/1 for an alternative implementation using a special label instead of boolean flags, which is less error-prone but breaks the user transparency in certain cases.
  - Fixes https://github.com/prometheus/prometheus/issues/11397

Example (this always fails, as `__name__` is being dropped by `count_over_time`):

```
count_over_time({__name__!=""}[1m])

=> Error executing query: vector cannot contain metrics with the same labelset
```

Before:

```
label_replace(count_over_time({__name__!=""}[1m]), "__name__", "rate_$1", "__name__", "(.+)")

=> Error executing query: vector cannot contain metrics with the same labelset
```

After:

```
label_replace(count_over_time({__name__!=""}[1m]), "__name__", "rate_$1", "__name__", "(.+)")

=> 
rate_go_gc_cycles_automatic_gc_cycles_total{instance="localhost:9090", job="prometheus"} 1
rate_go_gc_cycles_forced_gc_cycles_total{instance="localhost:9090", job="prometheus"} 1
...
```

<!--
    Please give your PR a title in the form "area: short description".  For example "tsdb: reduce disk usage by 95%"

    If your PR is to fix an issue, put "Fixes #issue-number" in the description.

    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --signoff flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

jcreixell committed:
POC3: Delay deletion of __name__ label

  - Instead of immediately deleting the label, this implementation
    extends the `Series` and `Sample` structs with a boolean flag
    indicating whether `__name__` should be deleted at the end of the
    evaluation.
  - This allows using `label_replace` to optionally preserve the label
    before it is deleted, which can also help with certain cases in
    which prematurely deleting the `__name__` label can lead to "metrics
    with same labelset" errors.
  - This implementation is more intrusive as [POC2](https://github.com/jcreixell/prometheus/pull/1), as it requires
    manually propagating the flag (and could be more error prone), but
    it is fully transparent to the user.
  - See https://github.com/prometheus/prometheus/issues/11397

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

jcreixell committed:
Fix lint issue

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

jcreixell committed:
Allow preserving __name__ via label_join

  - label_join can also be used to relabel __name__. When the target
    label is __name__, we should not delete in the last step.

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

jcreixell committed:
Add unit tests

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

bboreham commented on promql/value.go:
```diff
@@ -193,7 +194,8 @@ type Sample struct {
 
-	Metric labels.Labels
+	Metric         labels.Labels
+	ShouldDropName bool
```
8 bytes per sample seems too high an overhead for this feature.
I'm interested which operations cannot be done with the flag at Series level.

jcreixell replied:
I _think_ I may just need to refactor the `Vector` type to make it a struct instead of a `[]Sample`, and then move the flag to `Vector`. I will give it a try

jcreixell replied:
I spent some time today refactoring `Vector` to turn it into a struct in the form 

```go
type Vector struct {
	Samples        []Sample
	ShouldDropName bool
}
```

However, after reflecting on it, I don't think this change makes sense. For example, the following query may return a mix of results, some of which need the `__name__` label removed and others don't: 

```
rate(alertmanager_http_request_duration_seconds_count[1m]) or up{foo="bar"}
```

Another note: the `Sample` struct is only used for vectors, and each Sample contains a full metric, including labels (maybe it would be less confusing to call it `VectorSample`?)

```go
type Vector []Sample

type Sample struct {
	T int64
	F float64
	H *histogram.FloatHistogram

	Metric         labels.Labels
}
```

It is the equivalent to `Series` in the case of a `Matrix`:

```go
type Matrix []Series

type Series struct {
	Metric         labels.Labels `json:"metric"`
	Floats         []FPoint      `json:"values,omitempty"`
	Histograms     []HPoint      `json:"histograms,omitempty"`
}
```

IMHO, the drop flag needs to be a property of the metric, since it is unique to it, and in the case of vector, that means it belongs in the `Sample` struct, WDYT? Any suggestions on how to benchmark this?

bboreham replied:
What if we just drop `__name__` altogether in cases like `rate(alertmanager_http_request_duration_seconds_count[1m]) or up{foo="bar"}` ?

bboreham replied:
On second thoughts I don't think that is a good idea.

It sounds like `Vector` does need `ShouldDropName` on each series.  Is this the only place `Sample` is used?
If not, maybe `Sample` needs to split into two types?

jcreixell replied:
The only other place I found where `Sample` is used outside the `promql` package is [here](https://github.com/prometheus/prometheus/blob/main/web/api/v1/json_codec.go#L132-L137). Since it is a custom marshaler with handpicked fields, I don't think we need to add a new type.

bboreham commented on promql/engine_test.go:
```diff
@@ -3096,6 +3101,146 @@ func TestInstantQueryWithRangeVectorSelector(t *testing.T) {
+		expectedErr string
+		ts          time.Time
+	}{
+		"does not drop __name__ for instant queries": {
```
As far as I can see, all queries in this test are instant queries.  Did you mean 'vector selector' ?

bboreham commented on promql/engine_test.go:
```diff
@@ -3096,6 +3101,146 @@ func TestInstantQueryWithRangeVectorSelector(t *testing.T) {
+			},
+		},
+		"drops __name__ for functions": {
+			expr:        "count_over_time({__name__!=\"\"}[1m])",
```
Please also add an aggregation, e.g. `sum by (__name__) (rate({foo="bar"}[5m]))` suggested at https://github.com/prometheus/prometheus/issues/11397#issuecomment-1451998792

bboreham commented on promql/engine_test.go:
```diff
@@ -3096,6 +3101,146 @@ func TestInstantQueryWithRangeVectorSelector(t *testing.T) {
+			},
+		},
+		"allows preserving __name__ via label_replace": {
+			expr: "label_replace(count_over_time({__name__!=\"\"}[1m]), \"__name__\", \"$1\", \"__name__\", \"(.+)\")",
```
This struck me as very weird.  I guess it needs to work, but is it a degenerate case of something useful? Like
```
label_replace(rate({foo="bar"}[5m]), "__name__", "rate5m_of_$1", "__name__", "(.*)")
```
suggested at https://github.com/prometheus/prometheus/issues/11397#issuecomment-1451998792

jcreixell replied:
I used this because it was easier to test (run into some difficulties using the `rate` function in tests), but I will give it another try, as I agree it makes sense to test the expected use case

bboreham commented on promql/engine.go:
```diff
@@ -1073,6 +1073,28 @@ func (ev *evaluator) Eval(expr parser.Expr) (v parser.Value, ws annotations.Anno
+		if vec.ContainsSameLabelset() {
+			ev.errorf("vector cannot contain metrics with the same labelset")
+		}
+	}
```
Can this go in a separate function for readability?

bboreham reviewed (commented):
Generally this seems fine, but I would like to understand the need for the flag at Sample level.

jcreixell committed:
Move name cleanup code to a method for readability

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

beorn7 commented:
Sorry for getting to this late.

I now see the problem discussed earlier.

So far, I was approaching the problem from first principles, not remembering how the data is actually structured.

I agree that adding a field to `Sample` is too much of an overhead, and also overkill because we only need to decide about dropping the name for the whole series. However, just adding a bool to the `Series` type might also not be enough because that's not always used. (But if you can make it work, great.)

My "first principles" idea was to use what the `Metric` field in `Sample` is pointing to. However, that's not a normal struct, but `labels.Labels`, which happens to be the very type @bboreham has been working on heavily to optimize it. (There multiple alternative implementations you can choose from via build labels.)

As there are different implementations of `labels.Labels` anyway, we can do different things to track name dropping there. For the original not-so-optimized version, the naive approach would be to wrap the current slice into a struct, which then can hold the bool. But then we still have the overhead per sample, unless we also turn it into a pointer so that `Sample` would just point to it (which would actually make `Sample` itself more lightweight).

The `dedupelabels` implementation is already a struct, so we could add a bool there (but interestingly, the current implementation is not using a pointer, so the weight of `Sample` is actually increased if you compile with `dedupelabels`. 

The flat-string implementation `stringlabels` is actually easiest to change. We could just encode the flag in the first byte or something.

Wild night-time thoughts. I hope it makes a bit of sense.

(If you have read things after that sentence above, that was just pieces of thought that I forgot to delete. Please ignore if you still have them in your email or something.)

beorn7 commented:
A much more mundane thing: I think many of the quite complicated tests in engine_test.go can be converted into framework test in https://github.com/prometheus/prometheus/tree/main/promql/promqltest/testdata . (There are a bunch of tests there where we have plans to convert them. But we should also not add more tests there if they can be expressed equally well in the PromQL test framework.)

bboreham commented:
> just encode the flag in the first byte

This pattern would also be available in `-tags dedupelabels`.

While coupling `labels` to `promql` is a little distasteful, there is already a lot of hackery for the sake of efficiency.

beorn7 commented:
Had some direct discussions with @jcreixell . We were wondering how consistently `labels.Labels` instances are actually shared between samples for the same metric. The whole "avoiding overhead by tracking things in `labels.Labels`" only works if `labels.Labels` is usually shared. On the other hand, this means that we cannot just mutate `labels.Labels` to set the "remove name later" flag because setting this flag happens very often and it would create a copy of `labels.Labels` each time.
The same problem happens when removing the name label in the current state of things, but here we play the trick of just sub-slicing without the 1st element in the (usual) case where the name label comes first in the sorted `labels.Labels` slice. (Which is a bit weird on its own, as you could decrease the performance by having weird label names. Just think of the possibilities with the future UTF-8 label names. :smile: )
In any case, those tricks have to be kept in mind.

beorn7 commented:
Another outcome of our discussion: Let's make this PR work with the current approach and then benchmark it with Prombench to get an idea about the actual impact.

jcreixell committed:
Move tests to promql_test framework

  - Also adds some new tests for aggregations

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

jcreixell committed:
Ensure name dropping information is propagated by aggretionk operators

  - topk, bottomk were not properly preserving the flag

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

jcreixell committed:
Delay name dropping for vector-scalar operations

  - Seems like I had missed this case

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

jcreixell committed:
Improve tests

  - Test concrete results instead of using errors as a proxy for __name__
    label being dropped

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

jcreixell committed:
Remove redundant assignment for ShouldDropName flag

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

jcreixell committed:
Add test for aggregations

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

jcreixell commented:
I have done a few passes to make sure I didn't miss any place where the flag needs to be propagated and am reasonably satisfied with the current implementation. Happy to open a PR upstream so that we can benchmark it before deciding on next steps if you agree @bboreham @beorn7

jcreixell committed:
Add tests for binary comparison operators

Signed-off-by: Jorge Creixell <jcreixell@gmail.com>

jcreixell committed:
Merge branch 'main' into poc3
